backend/protocol_rpc/endpoints.py

```
@@ -684,6 +684,133 @@ def admin_upgrade_contract_code(
+                message_hash = Web3.keccak(
+                    b"cancel_transaction" + Web3.to_bytes(hexstr=transaction_hash)
+                )
+                message = encode_defunct(primitive=message_hash)
+                signer = Account.recover_message(message, signature=signature)
+
+                if signer.lower() != transaction.from_address.lower():
+                    raise JSONRPCError(
+                        code=-32000,
+                        message="Only transaction sender can cancel",
+                        data={"signer": signer, "sender": transaction.from_address},
+                    )
```

Ruff (0.15.2)

[warning] Abstract `raise` to an inner function

(TRY301)